Use spdlog for logging

Montellese opened the pull request:
This replaces our homemade logging implementation with [spdlog](https://github.com/gabime/spdlog) which is a header-only C++11 logging library which depends on the recently introduced [{fmt}](https://github.com/fmtlib/fmt) library (so we can use the exact same formatting style).

I only tested this on Windows and Linux so far. It should compile on Android but I didn't runtime test it. OSX/iOS is completely untested. Please take a look at the following platform-specific interface implementations:
* POXIS: https://github.com/xbmc/xbmc/commit/dccc02073c3ad9088b80348a6e44dc456155a9e3#diff-a862c7434b0614fad35e45406ab477d4
* Darwin: https://github.com/xbmc/xbmc/commit/dccc02073c3ad9088b80348a6e44dc456155a9e3#diff-616ffcf8daa5bad68c294971c149117b
* Android: https://github.com/xbmc/xbmc/commit/dccc02073c3ad9088b80348a6e44dc456155a9e3#diff-4de9b444ed50f7831abc986d45ef8fda
* Windows: https://github.com/xbmc/xbmc/commit/dccc02073c3ad9088b80348a6e44dc456155a9e3#diff-26243219725e225f0e126ab73069f522

For Ubuntu there is a `libspdlog-dev` package available for Xenial at http://packages.ubuntu.com/xenial/libdevel/libspdlog-dev but I wasn't able to figure out which version it actually is because it says 1.6 but that's not a valid release of spdlog. ~~But I adjusted the code to work with it nonetheless (had to work around a missing feature in the Ubuntu package).~~ Unfortunately it is not at all compatible with version 0.11.0 I used for this PR because it is missing several include files and some essential methods in the logger implementation.

## Motivation and Context
The advantage of a standard logging implementation is that we don't have to maintain our own implementation and they are well tested.
spdlog comes with a set of sinks that cover all of our use cases (console/file/adb logging/VS/... logging) and more that we might want to use in the future (file rotation, ...). Furthermore it supports named loggers which should make logging in the code much easier and also make the log file easier to read because every class/instance follows the same guidelines (right now every class logs their class name differently or not at all). As an example I've implemented a named logger for the `CWebServer` class in 9ed2953de94f50cc578cfad1af18b873c33817e7.

Using templates I also managed to get rid of the difference between `Log` and `LogW` which makes the interface of `CLog` easier to use.

I had to fix some log messages (all collected in 6f1e4a6fb0e29fb693270335f1084cb36d41a9b6) because they didn't match in type and therefore weren't type safe.

## How Has This Been Tested?
Local testing on windows.

## Types of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

## TODOs
- [ ] better solution for `gai_strerror()` as char/wchar_t
- [ ] fix pkgconfig version
- [ ] fixed width for thread id and log level
- [ ] find a solution for `LOGINFO` vs. `LOGNOTICE`
- [x] add `TARGET_WIN10` defines
- [x] get rid of unnecessary trailing newlines in log messages
- [x] fix crash in `CDVDDemuxClient`

Montellese committed:
[cmake/depends/win32] add spdlog

Montellese committed:
fix some log messages with custom types/wchar_t

Montellese committed:
[utils] CLog: replace custom logging implementation with spdlog

Montellese committed:
[network] CWebServer: make use of new logging functionality

Montellese commented:
jenkins build this please

Montellese commented:
jenkins build this please

Montellese commented:
Hm I just noticed that this is not the code with the workaround for linux (as mentioned in the PR description). Will have to see where that code went :-/

Montellese commented:
jenkins build this please

wsnipex commented on xbmc/utils/log.cpp:
```diff
@@ -18,199 +18,114 @@
-  // the log folder location is initialized in the CAdvancedSettings
-  // constructor and changed in CApplication::Create()
+  // the given path must not be empty
+  assert(!path.empty());
```
the compiler will remove this for release builds. Can we instead just use stderr in this case?

Montellese commented:
jenkins build this please

Montellese commented:
Let's try again. Linux and OSX/iOS should be fine now as well. jenkins build this please

Does anyone know what's wrong with the win32 jenkins slave? It says
```
C:\jenkins\slave\workspace\WIN-32\project\BuildDependencies\msys64\mingw32\bin\strip.exe: unable to rename 'libavcodec/x86/vp8dsp_loopfilter.o'; reason: Permission denied
library.mak:18: recipe for target 'libavcodec/x86/vp8dsp_loopfilter.o' failed
```

@Paxxi: do you have a better idea of solving https://github.com/xbmc/xbmc/pull/11833/commits/de4bfcd12bc5abcfc4a6cfa54b42b74ed5c3ee51#diff-51f117f980fb6b3965b3c706aa60f8e0R82 by using `_T()`? `gai_strerror()` returns a `wchar_t*` on Windows but a `char*` on everything else but the code is not windows specific. I had to manually define `_T()` for all non-windows platforms in https://github.com/xbmc/xbmc/pull/11833/commits/a47e802c956b0e50f381a23007eb363ca5f9157f#diff-6194d9c391444cab0e1322d1cc1fcfc6R31. This is pretty ugly.

@wsnipex: I removed the assert. If the path is empty we simply don't log to `kodi.log` (same behaviour as before) but continue to log to the other sinks (new behaviour).

MilhouseVH commented:
When building on Ubuntu, this PR now fails with:
```[100%] Building CXX object CMakeFiles/kodi.dir/xbmc/platform/posix/main.cpp.o
/usr/bin/ccache /usr/bin/c++   -DNPT_CONFIG_ENABLE_LOGGING -I/home/neil/projects/kodi.git -I/home/neil/projects/kodi.git/lib -I/home/neil/projects/kodi.git/lib/gtest/include -I/home/neil/projects/kodi.git/xbmc -I/home/neil/projects/kodi.git/xbmc/linux -I/home/neil/projects/kodi.git/xbmc/cores/VideoPlayer -I/home/neil/projects/kodi.git/build -I/usr/include/freetype2 -I/home/neil/projects/kodi.git/build/cpluff/include -I/home/neil/projects/kodi.git/build/libdvd/include -I/usr/local/include -I/usr/include/lzo -I/home/neil/projects/kodi.git/build/include -I/usr/include/libxml2 -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/samba-4.0 -I/usr/include/python2.7 -I/usr/include/libdrm -I/home/neil/projects/kodi.git/lib/libUPnP -I/home/neil/projects/kodi.git/lib/libUPnP/Platinum/Source/Core -I/home/neil/projects/kodi.git/lib/libUPnP/Platinum/Source/Platinum -I/home/neil/projects/kodi.git/lib/libUPnP/Platinum/Source/Devices/MediaConnect -I/home/neil/projects/kodi.git/lib/libUPnP/Platinum/Source/Devices/MediaRenderer -I/home/neil/projects/kodi.git/lib/libUPnP/Platinum/Source/Devices/MediaServer -I/home/neil/projects/kodi.git/lib/libUPnP/Platinum/Source/Extras -I/home/neil/projects/kodi.git/lib/libUPnP/Neptune/Source/Core -I/home/neil/projects/kodi.git/lib/libUPnP/Neptune/Source/System/Posix  -std=c++11 -O3 -DNDEBUG -s   -D_LINUX -DTARGET_POSIX -DTARGET_LINUX -DHAVE_SSE=1 -DHAVE_SSE2=1 -DHAVE_SSE3=1 -DHAVE_SSSE3=1 -DHAVE_SSE4_1=1 -D__STDC_CONSTANT_MACROS -D_FILE_DEFINED -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -DHAVE_STD__U16_STRING=1 -DHAVE_STD__U32_STRING=1 -DHAVE_CHAR16_T=1 -DHAVE_CHAR32_T=1 -DHAVE_STDINT_H=1 -DHAS_BUILTIN_SYNC_ADD_AND_FETCH=1 -DHAS_BUILTIN_SYNC_SUB_AND_FETCH=1 -DHAS_BUILTIN_SYNC_VAL_COMPARE_AND_SWAP=1 -DHAVE_INOTIFY=1 -DHAVE_POSIX_FADVISE=1 -DHAVE_LOCALTIME_R=1 -DHAVE_INTTYPES_H=1 -DSPDLOG_FMT_EXTERNAL -DSPDLOG_DEBUG_ON -I/usr/include/fribidi -DFFMPEG_VER_SHA=\"3.1.6-Krypton\" -DUSE_STATIC_FFMPEG=1 -DHAVE_LIBMICROHTTPD=1 -DHAVE_MYSQL=1 -DHAVE_LIBSSH=1 -DHAVE_LIBXSLT=1 -DHAVE_ALSA=1 -DUSE_ALSA=1 -DHAVE_LIBUDEV=1 -DHAVE_DBUS=1 -DHAVE_LIBSMBCLIENT=1 -DHAVE_LIBPULSE=1 -DHAS_PYTHON=1 -DHAVE_LIBBLURAY=1 -DHAVE_LIBPLIST=1 -DHAVE_LIBNFS=1 -DHAVE_OPENSSL=1 -DUSE_UPNP=1 -DHAVE_XBMC_NONFREE=1 -DHAS_DVD_DRIVE -DLIRC_DEVICE=\"/dev/lircd\" -DHAVE_LIRC=1 -DHAVE_LIBGL=1 -DHAVE_X11=1 -DHAVE_LIBXRANDR=1 -DHAVE_LIBEGL=1 -DBIN_INSTALL_PATH=\"/usr/local/lib/kodi\" -DINSTALL_PATH=\"/usr/local/share/kodi\" -std=c++11 -o CMakeFiles/kodi.dir/xbmc/platform/posix/main.cpp.o -c /home/neil/projects/kodi.git/xbmc/platform/posix/main.cpp
[100%] Linking CXX executable kodi.bin
/usr/bin/cmake -E cmake_link_script CMakeFiles/kodi.dir/link.txt --verbose=1
/usr/bin/ccache /home/neil/projects/kodi.git/build/ffmpeg-link-wrapper /usr/bin/c++  -std=c++11 -O3 -DNDEBUG -s   -fuse-ld=gold CMakeFiles/kodi.dir/xbmc/platform/posix/main.cpp.o -o kodi.bin -rdynamic -Wl,--whole-archive build/network/httprequesthandler/python/network_httprequesthandlers_python.a build/network/httprequesthandler/network_httprequesthandlers.a build/network/upnp/network_upnp.a build/upnp/libupnp.a build/build/swig/libpython_binding.a build/interfaces/python/python_interface.a build/interfaces/legacy/wsgi/legacy_interface_wsgi.a build/interfaces/legacy/legacy_interface.a build/rendering/gl/rendering_gl.a build/unrarxlib/libunrarxlib.a build/network/dacp/network_dacp.a build/cdrip/cdrip.a build/posix/platform_posix.a build/windowing/X11/windowing_X11.a build/filesystem/posix/filesystem_posix.a build/storage/linux/storage_linux.a build/powermanagement/linux/powermanagement_linux.a build/peripherals/bus/linux/peripherals_bus_linux.a build/network/linux/network_linux.a build/input/touch/generic/input_touch_generic.a build/input/touch/input_touch.a build/input/linux/input_linux.a build/linuxsupport/linuxsupport.a build/cores/VideoPlayer/videorenderers/shaders/videoshaders.a build/cores/VideoPlayer/videorenderers/videorenderers.a build/cores/VideoPlayer/process/process.a build/cores/VideoPlayer/subtitles/dvdsubtitles.a build/cores/VideoPlayer/inputstreams/dvdinputstreams.a build/cores/VideoPlayer/demuxers/dvddemuxers.a build/cores/VideoPlayer/codecs/video/dvdvideocodecs.a build/cores/VideoPlayer/codecs/overlay/dvdoverlaycodecs.a build/cores/VideoPlayer/codecs/audio/dvdaudiocodecs.a build/cores/VideoPlayer/codecs/dvdcodecs.a build/cores/VideoPlayer/VideoPlayer.a build/video/windows/video_windows.a build/video/jobs/video_jobs.a build/video/dialogs/video_dialogs.a build/video/video.a build/windows/windows.a build/windowing/windowing.a build/view/view.a build/utils/utils.a build/threads/threads.a build/storage/storage.a build/rendering/rendering.a build/programs/programs.a build/powermanagement/powermanagement.a build/playlists/playlists.a build/platform/platform_common.a build/pictures/pictures.a build/messagingHelpers/messagingHelpers.a build/messaging/messaging.a build/media/media.a build/listproviders/listproviders.a build/input/mouse/generic/input_mouse_generic.a build/input/mouse/input_mouse.a build/input/keyboard/generic/input_keyboard_generic.a build/input/keyboard/input_keyboard.a build/input/joysticks/generic/input_joystick_generic.a build/input/joysticks/dialogs/input_joystick_dialogs.a build/input/joysticks/input_joystick.a build/input/input.a build/guilib/guilib.a build/epg/epg.a build/dialogs/dialogs.a build/dbwrappers/dbwrappers.a build/commons/commons.a build/addonCallbacks_PVR/addonInterface_PVR.a build/addonCallbacks_Peripheral/addonInterface_Peripheral.a build/addonCallbacks_InputStream/addonInterface_InputStream.a build/addonCallbacks_GUI/addonInterface_GUI.a build/addonCallbacks_Game/addonCallbacks_Game.a build/addonCallbacks_Codec/addonInterface_Codec.a build/addonCallbacks_AudioEngine/addonInterface_AudioEngine.a build/addonCallbacks_AudioDSP/addonInterface_AudioDSP.a build/addonCallbacks_Addon/addonInterface_Addon.a build/addonsBinaryInterfaces/addonsBinaryInterfaces.a build/addons/addons.a build/xbmc/xbmc.a build/settings/windows/settings_windows.a build/settings/lib/settings_lib.a build/settings/dialogs/settings_dialogs.a build/settings/settings.a build/pvr/windows/pvr_windows.a build/pvr/timers/pvr_timers.a build/pvr/recordings/pvr_recordings.a build/pvr/dialogs/pvr_dialogs.a build/pvr/channels/pvr_channels.a build/pvr/addons/pvr_addons.a build/pvr/pvr.a build/profiles/windows/profiles_windows.a build/profiles/dialogs/profiles_dialogs.a build/profiles/profiles.a build/peripherals/dialogs/peripherals_dialogs.a build/peripherals/devices/peripherals_devices.a build/peripherals/bus/virtual/peripheral_bus_virtual.a build/peripherals/bus/peripherals_bus.a build/peripherals/addons/peripherals_addons.a build/peripherals/peripherals.a build/network/websocket/network_websockets.a build/network/network.a build/music/windows/music_windows.a build/music/tags/music_tags.a build/music/infoscanner/music_infoscanner.a build/music/dialogs/music_dialogs.a build/music/music.a build/interfaces/json-rpc/jsonrpc_interface.a build/interfaces/info/info_interface.a build/interfaces/generic/generic_interface.a build/interfaces/builtins/interfaces_builtins.a build/interfaces/interfaces.a build/games/windows/gameswindows.a build/games/tags/gametags.a build/games/ports/gameports.a build/games/dialogs/gamedialogs.a build/games/controllers/windows/games_controller_windows.a build/games/controllers/guicontrols/games_controller_guicontrols.a build/games/controllers/dialogs/games_controller_dialogs.a build/games/controllers/games_controller.a build/games/addons/savestates/gamesavestates.a build/games/addons/playback/gameplayback.a build/games/addons/gameaddons.a build/games/games.a build/filesystem/musicdatabase/musicdatabasedirectory.a build/filesystem/videodatabase/videodatabasedirectory.a build/filesystem/filesystem.a build/kissfft/kissfft.a build/events/windows/events_windows.a build/events/events.a build/cores/RetroPlayer/retroplayer.a build/cores/playercorefactory/playercorefactory.a build/cores/paplayer/paplayer.a build/cores/externalplayer/externalplayer.a build/cores/dll-loader/exports/util/exports_utils.a build/cores/dll-loader/exports/dllexports.a build/cores/dll-loader/dllloader.a build/cores/audioengine/audioengine.a build/cores/cores.a -Wl,--no-whole-archive libkodi.a -lpthread -lsqlite3 -lfreetype -lpcrecpp -lpcre build/libdvd/lib/libdvdnav.a build/libdvd/lib/libdvdread.a build/libdvd/lib/libdvdcss.a -ltinyxml -lyajl -lcdio /usr/local/lib/libfmt.a -llzo2 -lfribidi -ltag build/lib/libcrossguid.a -luuid -lz -lmicrohttpd -lmicrohttpd -lgcrypt -lgnutls -lp11-kit -lgmp -lhogweed -lgmp -lnettle -ltasn1 -lidn -lp11-kit -lz -lmysqlclient -lssh -lxslt -lxml2 -lasound -ludev -ldbus-1 -lsmbclient -lpulse -lpulse-mainloop-glib -lcap -lpython2.7 -lpthread -ldl -lutil -lpython2.7 -lssl -lcrypto -lGL -lGLU -lX11 -lXext -lXrandr -ldrm -lEGL build/network/httprequesthandler/python/network_httprequesthandlers_python.a build/network/httprequesthandler/network_httprequesthandlers.a build/network/upnp/network_upnp.a build/build/swig/libpython_binding.a build/interfaces/python/python_interface.a build/interfaces/legacy/wsgi/legacy_interface_wsgi.a build/interfaces/legacy/legacy_interface.a build/rendering/gl/rendering_gl.a build/unrarxlib/libunrarxlib.a build/network/dacp/network_dacp.a build/cdrip/cdrip.a build/posix/platform_posix.a build/windowing/X11/windowing_X11.a build/filesystem/posix/filesystem_posix.a build/storage/linux/storage_linux.a build/powermanagement/linux/powermanagement_linux.a build/peripherals/bus/linux/peripherals_bus_linux.a build/network/linux/network_linux.a build/input/touch/generic/input_touch_generic.a build/input/touch/input_touch.a build/input/linux/input_linux.a build/linuxsupport/linuxsupport.a build/cores/VideoPlayer/videorenderers/shaders/videoshaders.a build/cores/VideoPlayer/videorenderers/videorenderers.a build/cores/VideoPlayer/process/process.a build/cores/VideoPlayer/subtitles/dvdsubtitles.a build/cores/VideoPlayer/inputstreams/dvdinputstreams.a build/cores/VideoPlayer/demuxers/dvddemuxers.a build/cores/VideoPlayer/codecs/video/dvdvideocodecs.a build/cores/VideoPlayer/codecs/overlay/dvdoverlaycodecs.a build/cores/VideoPlayer/codecs/audio/dvdaudiocodecs.a build/cores/VideoPlayer/codecs/dvdcodecs.a build/cores/VideoPlayer/VideoPlayer.a build/video/windows/video_windows.a build/video/jobs/video_jobs.a build/video/dialogs/video_dialogs.a build/video/video.a build/windows/windows.a build/windowing/windowing.a build/view/view.a build/utils/utils.a build/threads/threads.a build/storage/storage.a build/rendering/rendering.a build/programs/programs.a build/powermanagement/powermanagement.a build/playlists/playlists.a build/platform/platform_common.a build/pictures/pictures.a build/messagingHelpers/messagingHelpers.a build/messaging/messaging.a build/media/media.a build/listproviders/listproviders.a build/input/mouse/generic/input_mouse_generic.a build/input/mouse/input_mouse.a build/input/keyboard/generic/input_keyboard_generic.a build/input/keyboard/input_keyboard.a build/input/joysticks/generic/input_joystick_generic.a build/input/joysticks/dialogs/input_joystick_dialogs.a build/input/joysticks/input_joystick.a build/input/input.a build/guilib/guilib.a build/epg/epg.a build/dialogs/dialogs.a build/dbwrappers/dbwrappers.a build/commons/commons.a build/addonCallbacks_PVR/addonInterface_PVR.a build/addonCallbacks_Peripheral/addonInterface_Peripheral.a build/addonCallbacks_InputStream/addonInterface_InputStream.a build/addonCallbacks_GUI/addonInterface_GUI.a build/addonCallbacks_Game/addonCallbacks_Game.a build/addonCallbacks_Codec/addonInterface_Codec.a build/addonCallbacks_AudioEngine/addonInterface_AudioEngine.a build/addonCallbacks_AudioDSP/addonInterface_AudioDSP.a build/addonCallbacks_Addon/addonInterface_Addon.a build/addonsBinaryInterfaces/addonsBinaryInterfaces.a build/addons/addons.a build/xbmc/xbmc.a build/settings/windows/settings_windows.a build/settings/lib/settings_lib.a build/settings/dialogs/settings_dialogs.a build/settings/settings.a build/pvr/windows/pvr_windows.a build/pvr/timers/pvr_timers.a build/pvr/recordings/pvr_recordings.a build/pvr/dialogs/pvr_dialogs.a build/pvr/channels/pvr_channels.a build/pvr/addons/pvr_addons.a build/pvr/pvr.a build/profiles/windows/profiles_windows.a build/profiles/dialogs/profiles_dialogs.a build/profiles/profiles.a build/peripherals/dialogs/peripherals_dialogs.a build/peripherals/devices/peripherals_devices.a build/peripherals/bus/virtual/peripheral_bus_virtual.a build/peripherals/bus/peripherals_bus.a build/peripherals/addons/peripherals_addons.a build/peripherals/peripherals.a build/network/websocket/network_websockets.a build/network/network.a build/music/windows/music_windows.a build/music/tags/music_tags.a build/music/infoscanner/music_infoscanner.a build/music/dialogs/music_dialogs.a build/music/music.a build/interfaces/json-rpc/jsonrpc_interface.a build/interfaces/info/info_interface.a build/interfaces/generic/generic_interface.a build/interfaces/builtins/interfaces_builtins.a build/interfaces/interfaces.a build/games/windows/gameswindows.a build/games/tags/gametags.a build/games/ports/gameports.a build/games/dialogs/gamedialogs.a build/games/controllers/windows/games_controller_windows.a build/games/controllers/guicontrols/games_controller_guicontrols.a build/games/controllers/dialogs/games_controller_dialogs.a build/games/controllers/games_controller.a build/games/addons/savestates/gamesavestates.a build/games/addons/playback/gameplayback.a build/games/addons/gameaddons.a build/games/games.a build/filesystem/musicdatabase/musicdatabasedirectory.a build/filesystem/videodatabase/videodatabasedirectory.a build/filesystem/filesystem.a build/upnp/libupnp.a build/kissfft/kissfft.a build/events/windows/events_windows.a build/events/events.a build/cores/RetroPlayer/retroplayer.a build/cores/playercorefactory/playercorefactory.a build/cores/paplayer/paplayer.a build/cores/externalplayer/externalplayer.a build/cores/dll-loader/exports/util/exports_utils.a build/cores/dll-loader/exports/dllexports.a build/cores/dll-loader/dllloader.a build/cores/audioengine/audioengine.a build/cores/cores.a -lgmp -lhogweed -lnettle -ltasn1 -lidn -lz -lmysqlclient -lssh -lxslt -lxml2 -lasound -ludev -ldbus-1 -lsmbclient -lpulse -lpulse-mainloop-glib -lcap -lpython2.7 -lpthread -ldl -lutil -lpython2.7 -lssl -lcrypto -lGL -lGLU -lX11 -lXext -lXrandr -ldrm -lEGL
build/games/addons/gameaddons.a(GameClient.cpp.o):GameClient.cpp:function GAME::CGameClient::NormalizeAudio(GAME::IGameAudioCallback*): error: undefined reference to 'GAME::CGameClientTiming::MAX_CORRECTION_FACTOR_PERCENT'
collect2: error: ld returned 1 exit status
CMakeFiles/kodi.dir/build.make:431: recipe for target 'kodi.bin' failed
make[2]: *** [kodi.bin] Error 1
make[2]: Leaving directory '/home/neil/projects/kodi.git'
CMakeFiles/Makefile2:506: recipe for target 'CMakeFiles/kodi.dir/all' failed
make[1]: *** [CMakeFiles/kodi.dir/all] Error 2
make[1]: Leaving directory '/home/neil/projects/kodi.git'
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
```

Montellese commented:
@MilhouseVH: thanks for the error. Looks like the issue is that `GAME::CGameClientTiming::MAX_CORRECTION_FACTOR_PERCENT` is a const integral which is inline initialized which is rejected by GCC when passed to a template method (see http://stackoverflow.com/questions/5391973/undefined-reference-to-static-const-int). I'll move the initialization to the CPP file and then it should be fine.

Paxxi commented:
`gai_strerror()` doesn't seem to be supported on UWP apps so I'm thinking the solution is creating a wrapper/helper method that calls either `gai_strerror()` or `WSAGetLastError()` and handles the W-utf8 conversion.

Either placing it on our socket class or I'm leaning towards adding a network header to platform/ and having it as a regular method there.

Will try to do the full review over the weekend but what I've seen so far looks good.

Montellese commented:
@Paxxi: when working on other stuff with WebSockets etc. I was already close before writing a wrapper around `gai_strerror()` and `WSAGetLastError()` but I never did it. But I'm also not sure where to best put it.

jenkins build this please

Montellese commented:
jenkins build this please

koying reviewed (approved):
Tested for droid, and basically give the same result so good for me.

3 remarks:
- Thread id is now the android thread id rather than the pthread one, which is actually nice :)
- CLog::MemDump seems gone, afaict. I personally use it often to debug, but can be re-added latter, I suppose.
- If there are changes in the way to use CLog, please be so kind as to detail them in the PR.

Thanks

MilhouseVH commented:
Before:
```
06:06:44.413 T:140196182277440  NOTICE: special://profile/ is mapped to: special://masterprofile/
06:06:44.413 T:140196182277440  NOTICE: -----------------------------------------------------------------------
06:06:44.413 T:140196182277440  NOTICE: Starting Kodi (18.0-ALPHA1 Git:4c792db). Platform: Linux x86 64-bit
06:06:44.413 T:140196182277440  NOTICE: Using Release Kodi x64 build
06:06:44.413 T:140196182277440  NOTICE: Kodi compiled Mar  2 2017 by GCC 6.2.0 for Linux x86 64-bit version 4.10.1 (264705)
06:06:44.413 T:140196182277440  NOTICE: Running on LibreELEC (Milhouse) - Version: devel-20170302210234-#0302-g6cc9ecb [Build #0302] 9.0, kernel: Linux x86 64-bit version 4.10.1
06:06:44.413 T:140196182277440  NOTICE: FFmpeg version/source: ffmpeg-3.1-kodi
06:06:44.413 T:140196182277440  NOTICE: Host CPU: Intel(R) Atom(TM) CPU D525 @ 1.80GHz, 4 cores available
06:06:44.414 T:140196182277440  NOTICE: special://xbmc/ is mapped to: /usr/share/kodi/
06:06:44.414 T:140196182277440  NOTICE: special://xbmcbin/ is mapped to: /usr/lib/kodi
```

After:
```
16:20:08.865986 T:883 info: special://profile/ is mapped to: special://masterprofile/
16:20:08.866141 T:883 info: -----------------------------------------------------------------------
16:20:08.866198 T:883 info: Starting Kodi (18.0-ALPHA1 Git:4c792db). Platform: Linux x86 64-bit
16:20:08.866239 T:883 info: Using Release Kodi x64 build
16:20:08.866285 T:883 info: Kodi compiled Mar 10 2017 by GCC 6.2.0 for Linux x86 64-bit version 4.10.1 (264705)
16:20:08.866532 T:883 info: Running on LibreELEC (Milhouse) - Version: devel-20170310160625-#0310d-g748c5dc [Build #0310d] 9.0, kernel: Linux x86 64-bit version 4.10.1
16:20:08.866600 T:883 info: FFmpeg version/source: ffmpeg-3.1-kodi
16:20:08.866638 T:883 info: Host CPU: Intel(R) Atom(TM) CPU D525 @ 1.80GHz, 4 cores available
16:20:08.866675 T:883 info: special://xbmc/ is mapped to: /usr/share/kodi/
16:20:08.866708 T:883 info: special://xbmcbin/ is mapped to: /usr/lib/kodi
```

The most obvious differences

1. increased time resolution
2. variable width thread id
3. variable width level (info, warning, etc.)

I'm sure others will disagree but 2 and 3 will result in a "messier" log, perhaps making it a little harder to "scan", for instance this:
```
16:20:18.953167 T:903 error: CRepository: failed read 'http://addons.libreelec.tv/9.0/Generic/x86_64/addons.xml.gz.md5'
16:20:18.968869 T:1007 info: initializing python engine.
16:20:18.976032 T:1006 info: initializing python engine.
16:20:18.977389 T:931 info: CPythonInvoker(7, /storage/.kodi/addons/script.openelec.devupdate/service.py): script successfully run
16:20:19.041967 T:903 info: easy_acquire - Created session to http://lrusak.libreelec.tv

16:20:19.089715 T:927 error: CCurlFile::Open failed with code 404 for http://raw.github.com/AwesomeStreams/AwesomeStreams/master/addons.xml.md5
16:20:19.090685 T:927 error: Open - failed to open source <http://raw.github.com/AwesomeStreams/AwesomeStreams/master/addons.xml.md5>
16:20:19.090795 T:927 error: CRepository: failed read 'http://raw.github.com/AwesomeStreams/AwesomeStreams/master/addons.xml.md5'
16:20:19.201335 T:1010 info: CFileCache::Process - Source read didn't return any data! Hit eof(?)
16:20:22.547502 T:931 info: Python script stopped
16:20:22.682816 T:921 info: script.grab.fanart: Grab Fanart Service Started
16:20:22.791179 T:925 info: CPythonInvoker(4, /storage/.kodi/addons/service.xbmc.versioncheck/service.py): script successfully run
```

And yes that is a blank line in the middle of the log - there are several more at the end here, including a partially written line: http://sprunge.us/OjiO

MilhouseVH commented:
Also, with `spdlog`, the kodi log entries are being written simultaneously to stdout as well as kodi.log.

Montellese commented:
@koying: Thanks for the review and feedback. Everything should still work the same except that you can now use the python formatting (same as with `StringUtils::Format()`) and that you can use named loggers but you don't have to (but it can be very nice). I can bring `MemDump` back. I grepped whether it was being used anywhere and since it wasn't I removed it. But if it's useful for debugging I can add it back and add a comment mentioning that fact.

@MilhouseVH: thanks for testing and your feedback. Concerning your points:
* is the higher time resolution good or bad? I can adjust it if required.
* I'll take a look at the variable width thread id and level
* I wasn't sure if the console output makes sense or not. That's why there's a TODO at https://github.com/xbmc/xbmc/pull/11833/files#diff-a862c7434b0614fad35e45406ab477d4R35. If we don't want this I just need to remove two lines of code in `xbmc/platform/posix/PosixInterfaceForCLog.cpp`.

koying commented:
@Montellese Don't bother MemDump. I'll put it back when I need it. Probably only useful on "some" platforms where using a real debugger is a misery ;)
- +1 to fixed width thread id, but it's minor
- +1 also for log not being output to stdout by default. I have some code that redirect stdout to logcat on droid, so that'd duplicate the logging.

MilhouseVH commented:
I prefer the higher time resolution.

My vote would be that we don't log to stdout as well as kodi.log, just as we don't log to stdout with the current logger.

stefansaraev commented:
sorry to hijack but

1. I see weird behaviour when doing "tail -f kodi.log" - last line is almost always cut at random character.
2. most, if not all binary would also need to be adapter tight? at least pvr.hts fails with unhandled exception in videoplayer thread

wsnipex commented:
+1 on either kodi.log or stdout (optimally configurable somehow)
re  last line is almost always cut at random character: sounds like buffered io instead of line flushing.
Thats usually good for performance on high traffic servers, but for kodi I'd prefer line flushed.

Montellese commented:
OK I've taken a closer look at the open issues:
* fixed width thread id and log level are not so easy to achieve but I'll take a look at extending `spdlog` because IMO this is necessary
* removed console logging from linux/android/osx
* added configuration that will flush after every log message
* most of the empty lines in the log seem to be coming from python add-ons. I'll take a closer look at what the problem might be
* @stefansaraev not sure why binary add-ons would need to be adjusted? where does the crash happen?

MilhouseVH commented:
> most of the empty lines in the log seem to be coming from python add-ons. I'll take a closer look at what the problem might be

Not just python - take a look at the "VDPAU Decoder capabilities" in this log: http://sprunge.us/WFOE (skip to 16:53:01.081579)

It looks like each line being logged has a newline (which really shouldn't be necessary): https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDCodecs/Video/VDPAU.cpp#L290-L291

However presumably the old logger automatically ignored these newlines, perhaps there's a way to do the same with spdlog?

MilhouseVH commented:
An example of the old logger and "VDPAU Decoder capabilities": http://sprunge.us/hPYh

MilhouseVH commented:
It also looks like there's an incorrect mapping between old and new log levels - `INFO` and `NOTICE` level messages are both displayed as `info` with `spdlog`.

old logger with debug enabled - NOTICE and INFO messages both present:
```
00:00:02.783 T:140637808098624  NOTICE: Enabled debug logging due to GUI setting (2)
...
00:00:02.801 T:140637808098624    INFO: creating subdirectories
00:00:02.801 T:140637808098624    INFO: userdata folder: special://masterprofile/
00:00:02.801 T:140637808098624    INFO: recording folder:
00:00:02.801 T:140637808098624    INFO: screenshots folder: /storage/screenshots/
00:00:02.812 T:140637808098624  NOTICE: Running database version Addons27
```

With the old logger, the INFO level messages would not be present in the log when debugging is disabled.

However with `spdlog` and debug disabled, both `NOTICE` and `INFO` are present and indistinguishable:
```
16:46:54.469473 T:1638 info: Disabled debug logging due to GUI setting. Level 0.
...
16:46:54.485795 T:1638 info: creating subdirectories
16:46:54.485897 T:1638 info: userdata folder: special://masterprofile/
16:46:54.485936 T:1638 info: recording folder: 
16:46:54.485973 T:1638 info: screenshots folder: /storage/screenshots/
16:46:54.498754 T:1638 info: Running database version Addons27
```

Montellese commented:
@MilhouseVH: spdlog doesn't have a `notice` level so I had to map it to `info`. I wasn't aware of `INFO` log messages not being in the log when debugging is disabled. I always thought it would only hide the `DEBUG` log messages.

I also tried to find the code that "handles" newlines in log messages but couldn't. There is generic code that replaces newlines with indented newlines and there is posix specific code which adds a newline after every written log message. So I'm not sure where those newlines are removed. IMO log messages shouldn't contain a newline at the end.

wsnipex commented on cmake/modules/FindSpdlog.cmake:
```diff
@@ -0,0 +1,42 @@
+#   Spdlog::Spdlog   - The Spdlog library
+
+if(PKG_CONFIG_FOUND)
+  pkg_check_modules(PC_SPDLOG spdlog QUIET)
```
since older versions floating around on linux dists don't seem to work, we have to specify a minimum version here

Paxxi commented on xbmc/platform/win32/Win32InterfaceForCLog.cpp:
```diff
@@ -20,21 +18,21 @@
 
-#include "utils/log.h"
-
 #ifndef TARGET_WINDOWS
```
should be `#if !defined(TARGET_WINDOWS) || !defined(TARGET_WIN10)` for now, might be needed to make a separate logger for uwp(the win10 target) later.

Montellese replied:
You mean `#if !defined(TARGET_WINDOWS) && !defined(TARGET_WIN10)` right?

Paxxi replied:
yep, you're correct, should be &&

Paxxi reviewed (approved):
As usual, nothing code-wise to complain about :)

Only the ifdefs for TARGET_WINDOWS that needs a check for TARGET_WIN10 as well for now.

stefansaraev commented:
@Montellese figured it out, it was not in pvr.hts

before spdlog:
http://sprunge.us/AOPY

after spdlog:
http://sprunge.us/YKVW

from wat I see, the unhandled exception seems to be exactly at the place where it previously logged:
```
DEBUG: CDVDDemuxClient::ParsePacket - {4} samplerate changed from 3 to 48000
```

this patch http://sprunge.us/hQea fixed it.

MilhouseVH commented:
It does seem that there's going to be a lot more "noise" with `spdlog` because it doesn't differentiate between NOTICE and INFO - a lot of debug-type items are being logged at INFO level which would not normally be visible with the old logger when debug is disabled, but will now be logged as standard with `spdlog`. It also means the NOTICE level is obsolete, which could be confusing when there is a lot of code still using LOGNOTICE.

Unfortunately I don't have a solution, just raising it as a concern.

Montellese commented:
@stefansaraev: Thanks for looking into this. That makes sense because the newly available python formatting style uses `{}` as delimiters so that will confuse the formatter.

@MilhouseVH: As mentioned before I wasn't aware of the difference between LOGINFO and LOGNOTICE. I never understood what the difference should be and I always used LOGINFO for stuff that was of general usefulness. But then again I always run with debug logging so I wouldn't notice the logs not being there.

Montellese committed:
remove unnecessary trailing newline from log messages

Montellese commented:
* added a commit to get rid of unnecessary trailing newlines in log messages
* fixed the crashes in `CDVDDemuxClient`
* added a TODO list to the PR description
* working on a fixed width patch for `spdlog`

MilhouseVH commented on xbmc/cores/DllLoader/DllLoader.cpp:
```diff
@@ -213,7 +213,7 @@ void DllLoader::PrintImportTable(ImportDirTable_t *ImportDirTable)
   int HavePrinted = 0;
 
-  CLog::Log(LOGDEBUG, "The Coff Image contains the following imports:\n\n");
+  CLog::Log(LOGDEBUG, "The Coff Image contains the following imports:\n");
```
Did you mean to keep `\n` here?

Montellese replied:
I always only removed the last `\n` because in some cases it looked like there's some kind of connection between multiple log messages.

MilhouseVH replied:
Yes, I guess that might have been the intention, but as we know that's now going to result in a blank line appearing in the log, which is going to be a little weird. If separation between log messages is required then shouldn't it log a blank message?

MilhouseVH commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecIMX.cpp:
```diff
@@ -270,7 +270,7 @@ bool CIMXCodec::VpuAlloc(VpuMemDesc *vpuMem)
   if (ret)
   {
-    CLog::Log(LOGERROR, "%s: vpu malloc frame buf of size %d failure: ret=%d \r\n",__FUNCTION__, vpuMem->nSize, ret);
+    CLog::Log(LOGERROR, "%s: vpu malloc frame buf of size %d failure: ret=%d \r",__FUNCTION__, vpuMem->nSize, ret);
```
Is the `\r` likely to cause a problem?

Montellese replied:
Nice spotting. That should probably go as well.

wsnipex commented:
I tried building on ubuntu and it fails with:
http://pastebin.com/SUU1h4MT

libfmt:3.0.0+ds-2~yakkety from PPA
spdlog: tried 0.11 and 0.12

wsnipex commented:
updating libfmt to 3.0.1 fixes my build issue. I'll update the PPA accordingly, but we need to require that version in cmake

Memphiz reviewed (approved):
Tested this on osx and ios and it looks like it works exactly like on the other platforms (+ the extra sink on osx that routes all logging into the "all messages" system log).

MartijnKaijser commented:
needs a rebase

Razzeee commented:
This is a really interesting find. We have one logtype more then most frameworks I'm aware of.
https://github.com/xbmc/xbmc/blob/master/xbmc/utils/log.cpp#L29
Has `notice` and `severe` that are odd (my POV) log levels, as far as I've seen those. On the other hand we're missing a `trace`.

For reference:
Common logging api for .net: https://cmatskas.com/an-introduction-to-common-logging-api-2/
NLog: https://github.com/NLog/NLog/wiki/Configuration-file#log-levels
SPD itself: https://github.com/gabime/spdlog/blob/13fb2550c68b3bb02bd7e422f9bd2523bf91c104/include/spdlog/sinks/syslog_sink.h#L36-L42

Montellese commented:
Yeah trace, debug, info, warning, error, fatal / critical are the log levels I'm used to as well from other projects like log4j (https://logging.apache.org/log4j/1.2/apidocs/org/apache/log4j/Level.html) or log4cxx (https://svn.apache.org/repos/asf/logging/site/trunk/docs/log4cxx/apidocs/classlog4cxx_1_1_level.html). The Poco framework also has a notice level though (https://pocoproject.org/docs/Poco.Logger.html#13781).

The main blocker right now is the missing fixed width for thread id and log level. I've made a hack patch to get it working but I didn't like it and therefore didn't get around to cleaning it up for upstream inclusion.

wsnipex commented:
notice is a very common log level, severe is common enough as well, but not as much

Razzeee commented:
We could just merge this now and later move to this when it hits spd? https://github.com/gabime/spdlog/pull/471

But doesn't solve the thread id I guess.

Razzeee commented:
I just realized that the fixed lenth for the thread id also is a "new feature". It's a welcome one, but is not the case in our current implementation.

Razzeee commented:
https://github.com/gabime/spdlog/pull/473
Has been merged.

Montellese commented:
See https://github.com/gabime/spdlog/pull/475. As soon as I've got feedback I'll rebase this.

Montellese committed:
[utils] CLog: use padded formatting for thread id and log level

Montellese commented:
This is what it would look like with my PR:
```
18:49:16.023799 T:12376  WARNING: CSettingsManager: missing version attribute
18:49:16.034800 T:12376     INFO: Default Video Player: VideoPlayer
18:49:16.034800 T:12376     INFO: Default Audio Player: paplayer
18:49:16.034800 T:12376     INFO: Disabled debug logging due to GUI setting. Level 1.
18:49:16.036800 T:12376     INFO: CMediaSourceSettings: loading media sources from special://masterprofile/sources.xml
18:49:16.081802 T:12376     INFO: Loading player core factory settings from special://xbmc/system/playercorefactory.xml.
18:49:16.101803 T:12376    DEBUG: CPlayerCoreConfig::<ctor>: created player VideoPlayer
```

Montellese commented:
I've rebased this but it's NOT READY for merge yet. It depends on my PR submitted to spdlog and will require new packages for win32 and depends.

Montellese commented:
So I've tried to get format padding support into spdlog (see https://github.com/gabime/spdlog/pull/475) but he didn't accept the idea in general. Since this is something we really want / need in Kodi there are basically two ways:

1. use spdlog with a custom patch
2. use another logging library with a wrapper for `libfmt` (because otherwise we will have two different formatting patterns for `StringUtils::Format()` and `CLog::Lo()`)

I won't have the time to evaluate another logging library (#2) but I would be able to update and extract the custom patch (#1).

What's it gonna be?

Paxxi commented:
carrying a custom patch isn't really an issue except for Linux as that places the support burden on wsnipex(not in theory but in reality).

Have we considered tweaking our own logging implementation to use variadic functions and fmt for formatting? That would likely be the simplest solution but we wouldn't gain any new features.

phunkyfish commented:
@Montellese is this something you still want to merge. Other options are close it or put it on the backburner.

Montellese commented:
I actually tried to pick this up again about a month ago since in the meantime `spdlog` has fixed the issue with the alignment. But I wasn't able to get it working on Ubuntu 16.04 because there we have to support a rather old version of `libspdlog-dev` which does not support this feature yet.

I also tried adding a custom version to depends but ran into issues with the compatibility with `libfmt`.

But once we drop support for Ubuntu 16.04 or we find another way around this issue I'm willing to pick this up again.

phunkyfish commented:
Cool, we have a backburner label now, so the idea is we set that and close it and then we can easily find it again. Sound ok?

garbear commented:
@Montellese maybe this helps: https://github.com/garbear/xbmc/commit/7fafa7c9c342abe2e66b984228a703182c57e0e7#diff-b7e7f0e8f4056134689b357baf3408ca

I had to update libfmt to v6: https://github.com/garbear/xbmc/commit/7fafa7c^

fuzzard commented:
We have had libfmt updated to 6.1.2 recently as well. May reduce some of the work for you.

https://github.com/xbmc/xbmc/commit/157b0ef05cfb99408a5f28472580ca1241f078f3

garbear commented:
it does, thanks!
